lib/coinbase/smart_contract.rb

```diff
@@ -2,257 +2,296 @@
+      end
+
+      private
+
+      # Converts a Solidity value to an appropriate Ruby type.
+      #
+      # @param solidity_value [Coinbase::Client::SolidityValue] The Solidity value to convert
+      # @return [Object] The converted Ruby value
+      # @raise [ArgumentError] If an unsupported Solidity type is encountered
+      #
+      # This method handles the following Solidity types:
+      # - Integers (uint8, uint16, uint32, uint64, uint128, uint256, int8, int16, int32, int64, int128, int256)
+      # - Address
+      # - String
+      # - Bytes (including fixed-size byte arrays)
+      # - Boolean
+      # - Array
+      # - Tuple (converted to a Hash)
+      #
+      # For complex types like arrays and tuples, the method recursively converts nested values.
+      def convert_solidity_value(solidity_value)
+        return nil if solidity_value.nil?
+
+        type = solidity_value.type
+        value = solidity_value.value
+        values = solidity_value.values
+
+        case type
+        when 'uint8', 'uint16', 'uint32', 'uint64', 'uint128', 'uint256',
+         'int8', 'int16', 'int32', 'int64', 'int128', 'int256'
+          value&.to_i
+        when 'address', 'string', /^bytes/
+          value
+        when 'bool'
+          if value.is_a?(String)
+            value == 'true'
+          else
+            !value.nil?
+          end
+        when 'array'
+          values ? values.map { |v| convert_solidity_value(v) } : []
+        when 'tuple'
+          if values
+            result = {}
+            values.each do |v|
+              raise ArgumentError, 'Error: Tuple value without a name' unless v.respond_to?(:name)
+
+              result[v.name] = convert_solidity_value(v)
+            end
+            result
+          else
+            {}
           end
-          result
         else
-          {}
+          raise ArgumentError, "Unsupported Solidity type: #{type}"
         end
-      else
-        raise ArgumentError, "Unsupported Solidity type: #{type}"
       end
-    end
-    private_class_method :convert_solidity_value
 
-    def self.contract_events_api
-      Coinbase::Client::ContractEventsApi.new(Coinbase.configuration.api_client)
-    end
-    private_class_method :contract_events_api
+      def contract_events_api
+        Coinbase::Client::ContractEventsApi.new(Coinbase.configuration.api_client)
+      end
 
-    def self.smart_contracts_api
-      Coinbase::Client::SmartContractsApi.new(Coinbase.configuration.api_client)
-    end
-    private_class_method :smart_contracts_api
-
-    def self.list_events_page(
-      network_id,
-      protocol_name,
-      contract_address,
-      contract_name,
-      event_name,
-      from_block_height,
-      to_block_height,
-      page
-    )
-      contract_events_api.list_contract_events(
-        Coinbase.normalize_network(network_id),
+      def smart_contracts_api
+        Coinbase::Client::SmartContractsApi.new(Coinbase.configuration.api_client)
+      end
```

Comment:
are we missing `list_smart_contracts ` function?

Reply:
Gracias

Reply:
There are a number of other APIs we'll need to add around this as well, namely "Get" for contracts not scoped to wallets.

But +1 we should add List in the time being.